cryostatio/cryostat-operator
# fix(cleanup): allow deletion to proceed without config map

# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
_______________________________________________

Fixes: #584 

## Description of the change:
The main change is to move the lock config map creation after handling the finalizer logic. The rest of the patch is additional test assurances that this doesn't break the name conflict handling introduced in #545.

## Motivation for the change:
This fixes the namespace deletion bug that's present in 2.3.0.

## How to manually test:
Ensure that deleting a conflicting CR doesn't delete/modify the original Cryostat installation:

1. Deploy operator into cryostat-operator-system
```
$ make deploy OPERATOR_IMG=quay.io/ebaron/cryostat-operator:namespace-del-fix-01
```
2. Create namespace c
```
$ kubectl create namespace c
```
3. Create Cryostat
```yaml
apiVersion: operator.cryostat.io/v1beta1
kind: Cryostat
metadata:
  name: cryostat-sample
  namespace: c
spec:
  minimal: false
  enableCertManager: true
  reportOptions:
    replicas: 0
```
4. Create ClusterCryostat
```yaml
apiVersion: operator.cryostat.io/v1beta1
kind: ClusterCryostat
metadata:
  name: cryostat-sample
spec:
  minimal: false
  enableCertManager: true
  installNamespace: c
  targetNamespaces:
  - c
  reportOptions:
    replicas: 0
```
5. Observe finalizer on ClusterCryostat
```
$ kubectl get clustercryostat cryostat-sample -o jsonpath='{.metadata.finalizers}'
["operator.cryostat.io/cryostat.finalizer"]
```
6. Delete ClusterCryostat
```
$ kubectl delete clustercryostat cryostat-sample
```
7. Check that ClusterRoleBinding and ConsoleLink were not deleted
```
$ kubectl get clusterrolebinding cryostat-22ca2798ad950a2953c096d484826501ddbae477d771250523c5523d9bd5a842
NAME                                                                        ROLE                                     AGE
cryostat-22ca2798ad950a2953c096d484826501ddbae477d771250523c5523d9bd5a842   ClusterRole/cryostat-operator-cryostat   46s
$ kubectl get consolelink cryostat-22ca2798ad950a2953c096d484826501ddbae477d771250523c5523d9bd5a842
NAME                                                                        TEXT       URL                                                                 MENU   AGE
cryostat-22ca2798ad950a2953c096d484826501ddbae477d771250523c5523d9bd5a842   Cryostat   https://cryostat-sample-cryostat-operator-system.apps.example.com          80s
```
8. Check that additionalCORSAllowedOrigins entry is still present in APIServer config:
```
$ kubectl get apiserver cluster -o jsonpath='{.spec.additionalCORSAllowedOrigins}'
["https://cryostat-sample-cryostat-operator-system\\.apps\\.example\\.com"]
```

Check that the namespace deletion bug is fixed:

1. Scale down the operator deployment
```
$ kubectl scale --replicas=0 deploy cryostat-operator-controller-manager
deployment.apps/cryostat-operator-controller-manager scaled
```
2. In another terminal, delete the namespace
```
$ kubectl delete namespace c
namespace "c" deleted
# hangs...
```
3. Scale the operator deployment back up
```
$ kubectl scale --replicas=1 deploy cryostat-operator-controller-manager
deployment.apps/cryostat-operator-controller-manager scaled
```
4. After some time, the namespace deletion proceeds
